---
title: "Fig1_WM_WGS_SomaticMutation"
author: "Sanghoon Lee"
format: html
editor: visual
---

## Figure 1. Summary of somatic mutations generated by Maftools visualization module.

A. An oncoplot depicting non-synonymous somatic alterations (SNVs and INDELs) for 11 mutated genes by decreasing frequency stratified by WM subtype

B. Barplots displaying distribution of mutations by variant classification in WM subtypes.

C. Barplots showing the number of patients with mutations in recurrent WM-genes.

## Lectures or useful websites

[Lecutre for oncoplots](https://bioconductor.org/packages/devel/bioc/vignettes/maftools/inst/doc/oncoplots.html)

[Oncoplots parameters](https://rdrr.io/bioc/maftools/man/oncoplot.html)

[Summarize, Analyze and Visualize MAF Files, various plot kinds](https://bioc.r-universe.dev/maftools/doc/maftools.html%20https://bioconductor.org/packages/devel/bioc/vignettes/maftools/inst/doc/maftools.html)

[Optimizing Oncoplot](https://www.bioconductor.org/packages/release/bioc/vignettes/maftools/inst/doc/oncoplots.html#04_Bar_plots)

## Step1. Load necessary R packages

```{r}
library(maftools) # bioconductor for oncoplots
library(data.table)
library(tidyverse)


## For complex heatmap
library(ComplexHeatmap)   # BiocManager::install("ComplexHeatmap")
library(circlize)  # cran
library(magick)  # cran

library(scales) ## To control decimal places in y-axis in ggplot.   label_number()
library(ggpubr) ## to use stat_compare_means() that calculates wilcoxon test
```

## Step2. Set working directory and input file names

```{r}
dir <- dirname(rstudioapi::getSourceEditorContext()$path); print(dir) #
setwd(dir); # "/Users/lees130/Library/CloudStorage/OneDrive-NYULangoneHealth/N05b_WaldenstromMacroglobulinemia_Dylan/12_Github_WMWGS_Quarto/Figure1_MAFPlot"

# ConsensusIndelSNV_MAFFile <- "/Users/lees130/Library/CloudStorage/OneDrive-NYULangoneHealth/N05b_WaldenstromMacroglobulinemia_Dylan/06e_AddSpotcheckMYD88_ToMaf/WM_SpotcheckCombined_SNVInde_Lite_ByManualMerge_235912row.maf.rds"

MAFData_All47s_Proc_OnlyWM11Gene_File <- "MAFData_All47s_Proc_OnlyWM11Gene.rds"

# WMSubtypeFile <- "/Users/lees130/Library/CloudStorage/OneDrive-NYULangoneHealth/N05b_WaldenstromMacroglobulinemia_Dylan/04a_MetadataCode_WMWGS_ByDylan_20250402/metadata/Metadata_WM_TreatmentSubtype_SH.txt"
WMSubtypeData_RmvLowQualityTumor_File <- "WMWGS_Consensus_MAFSomaticIndel_47s_ClinicalAnnotInput.txt"
```

## Step3. Read MAF data file and process it

```{r}
MAFData_All47s_Proc_OnlyWM11Gene <- readRDS(MAFData_All47s_Proc_OnlyWM11Gene_File)
dim(MAFData_All47s_Proc_OnlyWM11Gene); MAFData_All47s_Proc_OnlyWM11Gene[1:2,] # 113  16
#   Hugo_Symbol Chromosome Start_Position End_Position Reference_Allele Tumor_Seq_Allele2 Tumor_Sample_Barcode Matched_Norm_Sample_Barcode Tumor_Total_Read_Depth Tumor_Variant_Allele_Depth Tumor_Reference_Allele_Depth
# 1       MYD88       chr3       38141150     38141150                T                 C               01_076            01-076_T_S23_001                     32                         13                           19
# 2       CD79B      chr17       63929439     63929439                A                 G               01_076            01-076_T_S23_001                     24                         10                           14

table(MAFData_All47s_Proc_OnlyWM11Gene$Hugo_Symbol); table(MAFData_All47s_Proc_OnlyWM11Gene$Tumor_Sample_Barcode); length(table(MAFData_All47s_Proc_OnlyWM11Gene$Tumor_Sample_Barcode))
table(MAFData_All47s_Proc_OnlyWM11Gene$Variant_Classification); sum(table(MAFData_All47s_Proc_OnlyWM11Gene$Variant_Classification)) # 113 

# ### Exclude 3'Flank, 5'Flank, Intron, Slient, Spice_Region 
# MAFData_All47s_Proc_OnlyWM11Gene_ExcIntronSilent <- MAFData_All47s_Proc_OnlyWM11Gene %>% 
#                   dplyr::filter(grepl("Frame_Shift_Del|Frame_Shift_Ins|In_Frame_Del|Missense_Mutation|Nonsense_Mutation",  Variant_Classification))
# dim(MAFData_All47s_Proc_OnlyWM11Gene_ExcIntronSilent); table(MAFData_All47s_Proc_OnlyWM11Gene_ExcIntronSilent$Variant_Classification) # 86 16
# # Frame_Shift_Del   Frame_Shift_Ins      In_Frame_Del Nonsense_Mutation 
# # 9                 6                 1                 9 

###  after excluding Intron, Slient, Spice_Region, I have 6 cases with CD79B mutation.  CD79B mutation in WM9 is in Intron
unique(MAFData_All47s_Proc_OnlyWM11Gene$Tumor_Sample_Barcode[MAFData_All47s_Proc_OnlyWM11Gene$Hugo_Symbol=="CD79B"])
# [1] "01_076" "04_006" "WM74"   "WM82"   "WM85"  "WM9"   "WM90"  

table(MAFData_All47s_Proc_OnlyWM11Gene$Hugo_Symbol, MAFData_All47s_Proc_OnlyWM11Gene$Tumor_Sample_Barcode)
MAFData_All47s_Proc_OnlyWM11Gene[MAFData_All47s_Proc_OnlyWM11Gene$Hugo_Symbol=="TP53",]$Tumor_Sample_Barcode # [1] "WM33" "WM68" "WM70" "WM82" "WM84"


##### Let's add fake mutation genes for empty "Tumor_Sample_Barcode"
length(unique(MAFData_All47s_Proc_OnlyWM11Gene$Tumor_Sample_Barcode)) # 42   <<<--- 5 missing.
## Find which three is missing
unique(WMSubtypeData_RmvLowQualityTumor$Tumor_Sample_Barcode[!WMSubtypeData_RmvLowQualityTumor$Tumor_Sample_Barcode %in% unique(MAFData_All47s_Proc_OnlyWM11Gene$Tumor_Sample_Barcode) ])
# [1]  "WM49" "WM77" "WM80" "WM86" "WM97"

## Fake mutation gene row
FakeMutGene <- MAFData_All47s_Proc_OnlyWM11Gene[1,]
FakeMutGene_WM49 <- FakeMutGene; FakeMutGene_WM49$Hugo_Symbol <- "FakeGene"; FakeMutGene_WM49$Tumor_Sample_Barcode<-"WM49"
FakeMutGene_WM77 <- FakeMutGene; FakeMutGene_WM77$Hugo_Symbol <- "FakeGene"; FakeMutGene_WM77$Tumor_Sample_Barcode<-"WM77"
FakeMutGene_WM80 <- FakeMutGene; FakeMutGene_WM80$Hugo_Symbol <- "FakeGene"; FakeMutGene_WM80$Tumor_Sample_Barcode<-"WM80"
FakeMutGene_WM86 <- FakeMutGene; FakeMutGene_WM86$Hugo_Symbol <- "FakeGene"; FakeMutGene_WM86$Tumor_Sample_Barcode<-"WM86"
FakeMutGene_WM97 <- FakeMutGene; FakeMutGene_WM97$Hugo_Symbol <- "FakeGene"; FakeMutGene_WM97$Tumor_Sample_Barcode<-"WM97"

##### bind_rows with the fake MutGene Samples.
MAFData_All47s_IndelSNV_TopGene_WithFake <- dplyr::bind_rows(MAFData_All47s_Proc_OnlyWM11Gene, FakeMutGene_WM49) %>%     # FakeMutGene_WM22 %>% dplyr::bind_rows(FakeMutGene_WM27) %>%
                      dplyr::bind_rows(FakeMutGene_WM77) %>% dplyr::bind_rows(FakeMutGene_WM80) %>% dplyr::bind_rows(FakeMutGene_WM86) %>% dplyr::bind_rows(FakeMutGene_WM97)
length(unique(MAFData_All47s_IndelSNV_TopGene_WithFake$Tumor_Sample_Barcode)) # 47
dim(MAFData_All47s_IndelSNV_TopGene_WithFake) #118 16
saveRDS(MAFData_All47s_IndelSNV_TopGene_WithFake, "MAFData_All47s_IndelSNV_TopGene_WithFake.rds")

# ## Writing txt file.
MAFData_IndelSNV_HighFreqGene_OutFile <- "WMWGS_Consensus_MAF_IndelSNV_HighFreqGene_47s_ReadMAFMutInput.txt"
fwrite(MAFData_All47s_IndelSNV_TopGene_WithFake, file=MAFData_IndelSNV_HighFreqGene_OutFile, col.names=TRUE, row.names=FALSE, sep="\t", quote=FALSE)

```

## Step4. Make MAF object and oncoplot - Figure 1A

```{r}
## Checking VariantClassifcationNames   <<<<==============  Exclude "Splice_Region", "Silent", "Intron" "3'Flank", "5'Flank"
VariantClassficicationName_IndelSNV <- names(table(MAFData_All47s_Proc_OnlyWM11Gene$Variant_Classification));
VariantClassToRmv <- c("Splice_Region", "Silent", "Intron","3'Flank", "5'Flank") ## c() => Oncoplot_WMWGS_Consensus_MAFSomaticIndelSNV_WMSubtype_NoIGLHMUC_IncAllType_v1.7.pdf
VariantClassficicationName_IndelSNV_Proc <- setdiff(VariantClassficicationName_IndelSNV, VariantClassToRmv); length(VariantClassficicationName_IndelSNV_Proc)
print(VariantClassficicationName_IndelSNV_Proc)
# [1]      "Frame_Shift_Del"   "Frame_Shift_Ins"   "In_Frame_Del"     "Missense_Mutation" "Nonsense_Mutation"

## ++++++++++ Make MAF object
laml_object_IndelSNV = maftools::read.maf(maf = MAFData_IndelSNV_HighFreqGene_OutFile, clinicalData=WMSubtypeData_RmvLowQualityTumor_File,vc_nonSyn=VariantClassficicationName_IndelSNV_Proc, verbose = FALSE);
ClinicalData <- getClinicalData(x=laml_object_IndelSNV)
table(ClinicalData$WMSubtype); sum(table(ClinicalData$WMSubtype)) # 47
# G1_MBC_LIKE  G2_PC_LIKE G3_NotDetermined
# 10          22          15

## ++++++++++ Make Oncoplot
fabcolors <- c("Frame_Shift_Del"="lightgreen", "Frame_Shift_Ins"="darkgreen", "In_Frame_Del"="pink", "Missense_Mutation"="darkorange",  "Nonsense_Mutation"="red")

## WMSubtype Annotation
WMSubtypeCol <-  RColorBrewer::brewer.pal(n=length(table(ClinicalData$WMSubtype)),name = 'Paired')
names(WMSubtypeCol) = names(table(ClinicalData$WMSubtype))

DiagnosisSubtypeCol <-  RColorBrewer::brewer.pal(n=length(table(ClinicalData$Diagnosis)),name = 'Spectral')
names(DiagnosisSubtypeCol) = names(table(ClinicalData$Diagnosis))

Annot_cols = list(WMSubtype=WMSubtypeCol, Diagnosis=DiagnosisSubtypeCol)

oncoplot(maf = laml_object_IndelSNV, colors=fabcolors, clinicalFeatures=c('WMSubtype', 'Diagnosis'), sortByAnnotation=TRUE, annotationColor=Annot_cols)

### When you want to save the figure to PDF file. 
pdf(file="Fig1A_Oncoplot_WMWGS_Consensus_MAFSomaticIndelSNV_WMSubtype_NoIGLHMUC_NoIntronSilent_v1.0.pdf", width=9, height=7)
    par(mfrow=c(1,1))
    # oncoplot(maf = laml_object_IndelSNV, colors=fabcolors, clinicalFeatures='WMSubtype', sortByAnnotation=TRUE, annotationColor=WMSubtypeColors_Indel)
    oncoplot(maf = laml_object_IndelSNV, colors=fabcolors, clinicalFeatures=c('WMSubtype', 'Diagnosis'), sortByAnnotation=TRUE, annotationColor=Annot_cols)
           #  clinicalFeatures = c('MBC_LIKE','PC_LIKE','NotDetermined')) # clinicalFeatures='WMSubtype'
    ## Error - solution run "dev.off()"
    # Error in .External.graphics(C_layout, num.rows, num.cols, mat, as.integer(num.figures),  :  invalid graphics state # solution: dev.off()
dev.off()

```

## Step5. Split "MAFData_All47s_Proc_OnlyWM11Gene" to MBC-like, PC-like, or NotDetermined.

```{r}
## ++++++++++++ Read clinical data 
WMSubtypeData_RmvLowQualityTumor <- fread(WMSubtypeData_RmvLowQualityTumor_File, header=TRUE, stringsAsFactors=FALSE); 
dim(WMSubtypeData_RmvLowQualityTumor); WMSubtypeData_RmvLowQualityTumor[1:2,] # 47 4
#    Tumor_Sample_Barcode Diagnosis Treatment   WMSubtype
#                  <char>    <char>    <char>      <char>
# 1:               01_076        ND Rituximab G1_MBC_LIKE
# 2:               01_115        ND  NotDetermined  G2_PC_LIKE


## 1) MBC_Like  
MAFData_All47s_Proc_MBCLike <- MAFData_All47s_IndelSNV_TopGene_WithFake %>% dplyr::filter(Tumor_Sample_Barcode %in%
                                  WMSubtypeData_RmvLowQualityTumor$Tumor_Sample_Barcode[WMSubtypeData_RmvLowQualityTumor$WMSubtype=="G1_MBC_LIKE"])  %>% 
                                  dplyr::filter(grepl("Frame_Shift_Del|Frame_Shift_Ins|In_Frame_Del|Missense_Mutation|Nonsense_Mutation",  Variant_Classification)) ### Exclude Intron, Slient, 3'/5'Flank 
table(MAFData_All47s_Proc_MBCLike$Variant_Classification)
# Frame_Shift_Del   Frame_Shift_Ins      In_Frame_Del Missense_Mutation Nonsense_Mutation 
# 5                 2                 1                18                 4 

table(WMSubtypeData_RmvLowQualityTumor$Tumor_Sample_Barcode[WMSubtypeData_RmvLowQualityTumor$WMSubtype=="G1_MBC_LIKE"]) # 01_076 01_190   WM31   WM33   WM46   WM47   WM52   WM65   WM70    WM9
dim(MAFData_All47s_Proc_MBCLike); table(MAFData_All47s_Proc_MBCLike$Tumor_Sample_Barcode) # 30 16  by all genes.         # 41 16 by WM interesting genes
## # 01_076 01_190   WM31   WM33   WM46   WM47   WM52   WM65   WM70    WM9
##########   ++++++++++++++     Writing txt file.   - With all available genes.   DEL, INS, SNV
MAFData_IndelSNV_MBCLike_OutFile <- "WMWGS_Consensus_MAF_DELINSSNV_MBCLike10s_ReadMAFMutInput.txt"
fwrite(MAFData_All47s_Proc_MBCLike, file=MAFData_IndelSNV_MBCLike_OutFile, col.names=TRUE, row.names=FALSE, sep="\t", quote=FALSE)
saveRDS(MAFData_All47s_Proc_MBCLike, "MAFData_All47s_Proc_MBCLike.rds")

## 3) PC_Like
MAFData_All47s_Proc_PCLike <- MAFData_All47s_IndelSNV_TopGene_WithFake %>% dplyr::filter(Tumor_Sample_Barcode %in%
                                 WMSubtypeData_RmvLowQualityTumor$Tumor_Sample_Barcode[WMSubtypeData_RmvLowQualityTumor$WMSubtype=="G2_PC_LIKE"], Hugo_Symbol != "FakeGene") %>% 
                                 dplyr::filter(grepl("Frame_Shift_Del|Frame_Shift_Ins|In_Frame_Del|Missense_Mutation|Nonsense_Mutation",  Variant_Classification)) ### Exclude Intron, Slient, 3'/5'Flank 
table(WMSubtypeData_RmvLowQualityTumor$Tumor_Sample_Barcode[WMSubtypeData_RmvLowQualityTumor$WMSubtype=="G2_PC_LIKE"]) # 01_115 01_131 01_163 04_006   WM22   WM26   WM27   WM29   WM30   WM32   WM37   WM42   WM43   WM48   WM49   WM54   WM66   WM68   WM69   WM72   WM73    WM8
table(MAFData_All47s_Proc_PCLike$Variant_Classification)
# Frame_Shift_Del   Frame_Shift_Ins Missense_Mutation Nonsense_Mutation 
# 3                 3                25                 1 

dim(MAFData_All47s_Proc_PCLike); table(MAFData_All47s_Proc_PCLike$Tumor_Sample_Barcode) # 37784   16  by all genes    # 41 16 by WM insteresing genes.
## # 01_115 01_131 01_163 04_006   WM22   WM26   WM27   WM29   WM30   WM32   WM37   WM42   WM43   WM48   WM49   WM54   WM66   WM68   WM69   WM72   WM73    WM8
##########   ++++++++++++++     Writing txt file.   - With all available genes.   DEL, INS, SNV
MAFData_IndelSNV_PCLike_OutFile <- "WMWGS_Consensus_MAF_DELINSSNV_PCLike22s_ReadMAFMutInput.txt"
fwrite(MAFData_All47s_Proc_PCLike, file=MAFData_IndelSNV_PCLike_OutFile, col.names=TRUE, row.names=FALSE, sep="\t", quote=FALSE)
saveRDS(MAFData_All47s_Proc_PCLike, "MAFData_All47s_Proc_PCLike.rds")

## 4) NotDetermined
MAFData_All47s_Proc_NotDetermined <- MAFData_All47s_IndelSNV_TopGene_WithFake %>% dplyr::filter(Tumor_Sample_Barcode %in%
                              WMSubtypeData_RmvLowQualityTumor$Tumor_Sample_Barcode[WMSubtypeData_RmvLowQualityTumor$WMSubtype=="G3_NotAvail"], Hugo_Symbol != "FakeGene") %>% 
                              dplyr::filter(grepl("Frame_Shift_Del|Frame_Shift_Ins|In_Frame_Del|Missense_Mutation|Nonsense_Mutation",  Variant_Classification)) ### Exclude Intron, Slient, 3'/5'Flank 
table(WMSubtypeData_RmvLowQualityTumor$Tumor_Sample_Barcode[WMSubtypeData_RmvLowQualityTumor$WMSubtype=="G3_NotAvail"]) # 01_115 01_131 01_163 04_006   WM22   WM26   WM27   WM29   WM30   WM32   WM37   WM42   WM43   WM48   WM49   WM54   WM66   WM68   WM69   WM72   WM73    WM8
table(MAFData_All47s_Proc_NotDetermined$Variant_Classification)
# Frame_Shift_Del   Frame_Shift_Ins Missense_Mutation Nonsense_Mutation 
#              1                 1                18                 4 

dim(MAFData_All47s_Proc_NotDetermined); table(MAFData_All47s_Proc_NotDetermined$Tumor_Sample_Barcode) #   # 24 16 by WM insteresing genes.
## # WM67 WM74 WM77 WM80 WM82 WM83 WM84 WM85 WM86 WM87 WM89 WM90 WM95 WM96 WM97
##########   ++++++++++++++     Writing txt file.   - With all available genes.   DEL, INS, SNV
MAFData_IndelSNV_NotDetermined_OutFile <- "WMWGS_Consensus_MAF_DELINSSNV_NotDetermined11s_ReadMAFMutInput.txt"
fwrite(MAFData_All47s_Proc_NotDetermined, file=MAFData_IndelSNV_NotDetermined_OutFile, col.names=TRUE, row.names=FALSE, sep="\t", quote=FALSE)
saveRDS(MAFData_All47s_Proc_NotDetermined, "MAFData_All47s_Proc_NotDetermined.rds")

```

## Step6. Summary plot - Figure 1B and C

```{r}

## 1) MBC-Like for 10 smp  - MBCLike
laml_MAFData_IndelSNV_MBCLike = maftools::read.maf(maf = MAFData_IndelSNV_MBCLike_OutFile, clinicalData = WMSubtypeData_RmvLowQualityTumor_File)  
# Error: could not find function "validateMaf" ## I need "maftools::read.maff"

###### +++++++++++  Plot Summary in both Indel and SNV  +++++++++ #################
pdf(file="Fig1BC_Summaryplot_WMWGS_Consensus_MAFSomaticIndelSNV_NoIGLHMUC_MBC10s_v1.0.pdf", width=9, height=7)
      par(mfrow=c(1,1))
      plotmafSummary(maf=laml_MAFData_IndelSNV_MBCLike, rmOutlier = TRUE, addStat = 'median', dashboard = TRUE, titvRaw = FALSE)
dev.off()
##########################################################################################

## 2) PC-Like for 10 smp  - PCLike
laml_MAFData_IndelSNV_PCLike = maftools::read.maf(maf = MAFData_IndelSNV_PCLike_OutFile, clinicalData = WMSubtypeData_RmvLowQualityTumor_File)
###### +++++++++++  Plot Summary in both Indel and SNV  +++++++++ #################
pdf(file="Fig1BC_Summaryplot_WMWGS_Consensus_MAFSomaticIndelSNV_NoIGLHMUC_PC22s_v1.0.pdf", width=9, height=7)
    par(mfrow=c(1,1))
    plotmafSummary(maf=laml_MAFData_IndelSNV_PCLike, rmOutlier = TRUE, addStat = 'median', dashboard = TRUE, titvRaw = FALSE)
dev.off()
##########################################################################################

## 3) NotAvail for 11 smp  - NotAvail
laml_MAFData_IndelSNV_NotAvail = maftools::read.maf(maf = MAFData_IndelSNV_NotDetermined_OutFile, clinicalData = WMSubtypeData_RmvLowQualityTumor_File)
###### +++++++++++  Plot Summary in both Indel and SNV  +++++++++ #################
pdf(file="Fig1BC_Summaryplot_WMWGS_Consensus_MAFSomaticIndelSNV_NoIGLHMUC_NotDetermined5s_RmvInt3Flank5Flank_v1.0.pdf", width=9, height=7)
par(mfrow=c(1,1))
plotmafSummary(maf=laml_MAFData_IndelSNV_NotAvail, rmOutlier = TRUE, addStat = 'median', dashboard = TRUE, titvRaw = FALSE)
dev.off()
##########################################################################################


```
